README.md
# FacebookConnector plugin for Craft CMS 3.x
[![Codacy Badge](https://api.codacy.com/project/badge/Grade/a7f189dbcb6645f6bef11d3ad00b617f)](https://www.codacy.com/app/boscho87/facebook-connector?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=boscho87/facebook-connector&amp;utm_campaign=Badge_Grade)
[![Build Status](https://travis-ci.org/boscho87/facebook-connector.svg?branch=version-1.x.x)](https://travis-ci.org/boscho87/facebook-connector)
[![Codacy Badge](https://api.codacy.com/project/badge/Coverage/a7f189dbcb6645f6bef11d3ad00b617f)](https://www.codacy.com/app/boscho87/facebook-connector?utm_source=github.com&utm_medium=referral&utm_content=boscho87/facebook-connector&utm_campaign=Badge_Coverage)

Connect your Website with a Facebookpage

![Screenshot](resources/img/icon.png)

## FacebookConnector Overview
 ### Post your Craft entry as a Post on Facebook with simple Setup

## Requirements
This plugin requires Craft CMS 3.0.0-beta.23 or later.

## Contribute to the Project
if you want to help with this project, [read how to contribute](CONTRIBUTE.md)

## Installation
To install the plugin, follow these instructions.
1. Open your terminal and go to your Craft project:

        cd /path/to/project
        
2. Then tell Composer to load the plugin:

       composer require itscoding/facebook-connector
       
3. In the Control Panel, go to Settings → Plugins and click the “Install” button for FacebookConnector.

4. Configure the plugin

## Configuring FacebookConnector
### Post an Entry on facebook
add the file fieldconfig.php and return a callable that returns an array with your fields
`/path/to/project/fieldconfig.php`

```php
<?php
use itscoding\facebookconnector\FacebookConnector;
use craft\elements\Entry;

return function (Entry $entry) {
   
    //you can do some fancy stuff with callables 👍 !
    $description = function() use ($entry){
      return $entry->teaserSubTitle;  
    };
    
    //callables will not work here!
    return [
        //if true, the message is posted {default:true}
        'post_on_facebook' => $entry->post_on_facebook,
        //official doc https://developers.facebook.com/docs/graph-api/reference/v2.10/post
        //craft field to get the message to post from {default:''}
        'message' => $entry->message,
        //link to the entry page {default:$entry->getUrl()}
        'link' => $entry->getUrl(),
        //field to get the img url from {default:''} --> no image
        'picture' => (count($entry->fb_image) > 0) ? Craft::$app->request->getHostInfo() . $entry->fb_image->first()->getUrl() : '',
        //field to get the caption from {default:''}
        'caption' => $entry->getAuthor()->getName(),
        //field to get the description from {default:''}
        'description' => $description() 
    ];
};
```

### Fetch Entries from facebook
To fetch entries from Facebook, you need to run some scripts (if you need to reload posts automatic, setup cronjobs for this)

```shell
#in your project (craft) dir run:
php craft facebook-connector/entry-fetch/fetch-list #loads a list with the entries (id, title and content)
php craft facebook-connector/entry-fetch/fetch-detail #loads the details for all entries, saved before
```
be aware, that the entries are incomplete, if the details are missing! (you can check in your twig template if an entry is incomplete, before you render it)

##### Entry properties:
```php
  [
    'id' => $this->primaryKey(),  // like every craft entry
    'fbId' => $this->string(255)->unique(),
    'dateCreated' => $this->dateTime()->notNull(), // like every craft entry
    'dateUpdated' => $this->dateTime()->notNull(), // like every craft entry
    'created' => $this->string(255)->notNull(), 
    'has_detail' => $this->boolean(), // if the entry details are loaded 
    'type' => $this->string(), // type like ['share','photo','youtube']
    'image_src' => $this->string(510), //
    'image_height' => $this->string(), //
    'image_width' => $this->string(), // 
    'target' => $this->string(510), // target url from facebook 
    'url' => $this->string(510), // facebook post url
    'internal' => $this->boolean(), // true if the post is from one of the craft pages where the plugin is installed
    'uid' => $this->uid(), // like every craft entry
    'title' => $this->string(510), // should be clear
    'video' => $this->string(510), // [its a link] only set if its a video 'type' (atm. works only wiht youtube
    'content' => $this->text() // text body
    //event stuff (only when type  = event) 
    'start_time' => $this->string(45),
    'end_time' => $this->string(45),
    'event_cover_source' => $this->string(255),
    'event_cover_offset_x' => $this->string(45),
    'event_cover_offset_y' => $this->string(45),
  ]
```

##### Example (in action)
```html
<div class="col-lg-8 col-md-10 mx-auto">
    {% for fbEntry in craft.facebook.getEntries.where('internal = 0').andWhere('type != "event"').limit(3).all %}
        <div style="position: relative;">
            {% if fbEntry.type == 'youtube' %}
                <h3>    {{ fbEntry.title }}    </h3>
                <iframe width="560" height="315"
                        src="https://www.youtube.com/embed{{ fbEntry.video }}?rel=0&amp;controls=0&amp;showinfo=0"
                        frameborder="0"
                        allowfullscreen></iframe>
            {% else %}
                <div class="post-preview" style="float: left;width: 70%;">
                    <h2 class="post-title">
                        {{ fbEntry.title ?: 'Beitrag vom '~fbEntry.created|date('d.m.y') }}
                    </h2>
                    <p style="padding-bottom: 20px">{{ fbEntry.content }}</p>
                    <p class="post-meta" style="bottom: 0;position: absolute; ">
                        Gepostet auf
                        {% if fbEntry.target %}
                            <a href="{{ fbEntry.target }}" target="_blank"> Facebook</a>
                        {% else %}
                            Facebook
                        {% endif %}
                        {# decode the content, because it was encoded with json to stroe to more types of db's #}
                        am {{ fbEntry.created|date('d.m.y') }} um {{ {{ craft.facebook.decode(fbEntry.content) }} }}
                    </p>
                </div>
                <div style="float: left;width: 30%">
                    {% if fbEntry.image_src %}
                        <img class="img-fluid" src="{{ fbEntry.image_src }}" alt="">
                    {% endif %}
                </div>
            {% endif %}
            <div style="clear: both"></div>
            <hr>
            </div>
    {% endfor %}
</div>
```





Because i'm too lazy (I do not like writing Tutorials) [here is a youtube link](https://link.com), where the whole setup is explained
     
     feel free to Contribute a written description :)
 
## FacebookConnector Roadmap
Some things to do:

* Finish the Documentation
* Add more features [Ideas]-[priority 1=high 5=low] 
  
    - Add possibility to choose user to post - 5
    - Show Likes(and like infos) to an entry on the Website - 4
    - etc. (contact me if you have an idea) 

Brought to you by [Simon Müller itsCoding](https://www.itscoding.ch)


CHANGELOG.md
# FacebookConnector Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) and this project adheres to [Semantic Versioning](http://semver.org/).

#### Unreleased

### 1.4.0 - 2019-04-16
- Fixed Issue in Entry Parser with makes Fetching fail with content less then 25 Chars.

### 1.3.2 - 2019-04-16
- Fixed an Issue with the Entry Parser, because it fails and did not Load images.

### 1.3.1 - 2019-01-01
- Fixed Issue in Entry Parser with makes Fetching fail with content less then 25 Chars.

### 1.3.0 - 2019-01-01

## Update
- Default api is now v3.1
- The Titles now encoded with json before stored to the database
- Improved Content Parsing, all Emojis now will be replaced

####  Released
## 1.2.5 - 2018-01-20
### Fixed
- Fixed a bug in EntryPoster that throws an exception when saving an entry

## 1.2.4 - 2018-01-18
### Added
 - Added CI configuration to the Project
 - Dummy PHPUnit test
  
### Changed
 - removed v from the version numbers
 
### Fixed
 - Fixed an issue with the configLoader   


## 1.2.3 - 2017-12-22
### Changed
 - updated composer.json

## 1.2.0 - 2017-12-12
### Changed,
 - going to the first Stable Major Vesion
 
 
## 1.1.0-rc.2 - 2017-12-09
### Fixed
 - Fixed a bug where shared events throws exception on requesting the event details, created new type `shared_event` for this case


## 1.1.0-rc.1 - 2017-12-07
### Added
 - Event Deatail Data now also available 
### Updated
 - Updated Readme.md    

## 1.0.0-rc.1
## Added
 - Updated readme, fixed code issues

## 0.18.0 - 2017-11-05
### Added
 - Entries can now be Loaded from Facebook
 - Added a craft variable to get the entries from the database
 - Entries now "fechtable" with quries (like standard craft contents)
    
## 0.17.0 - 2017-11-02
 - Started with Implementation of "fetch Facebook Feed" 

## 0.16.0 - 2017-10-08
### Added
- Contribute Text
- Added Link to the Youtube Configuration description

### Fixed
- Updated the Documentation how its possible to use functions in `fieldconfig.php` #13
  
### Changed
- Refactored the EntryPoster  
  
## 0.15.0 - 2017-10-08
### Changed
- Readme Update

## 0.14.0 - 2017-09-25
### Added 
- First german Translations
- Added some lines of Doc

## 0.13.0 - 2017-09-25
### Fixed
- fixed codacy issues

## 0.12.0 - 2017-09-25
### Changed
- renamed vendor of the project

## 0.11.0 - 2017-09-25
### Changed
- Moved the token callback function to the dashboard page, so no more templates are needed

## 0.11.0 - 2017-09-25
### Added
- Boilerplate code added for the widget

## 0.10.0 - 2017-09-25
### Added
- Entry field options can no be overwritten in the file fieldconfig.php
  
## 0.9.0 - 2017-09-24
### Added
- Options can now be defined in the plugin settings

## 0.8.0 - 2017-09-24
### Changed
 - Changed the readme and added som information's about the project

## 0.7.0 - 2017-09-22
### Added
- Implemented OAuth Login
- Implemented post to facebook
- Implemented update a post on facebook

## 0.1.0 - 2017-09-16 (Initial Release)
### Added
 - Initial release


src/migrations/Install.php
<?php
/**
 * FacebookConnector plugin for Craft CMS 3.x
 *
 * @link      www.itscoding.ch
 * @copyright Copyright (c) 2017 boscho87\itscoding
 */

namespace itscoding\facebookconnector\migrations;

use Craft;
use craft\db\Migration;

/**
 * FacebookConnector Install Migration
 *
 * @author    boscho87\itscoding
 * @package   FacebookConnector
 * @since     0.1.0
 */
class Install extends Migration
{

    /**
     * @var string The database driver to use
     */
    public $driver;

    /**
     * @return boolean return a false value to indicate the migration fails
     * and should not proceed further. All other return values mean the migration succeeds.
     */
    public function safeUp()
    {
        $this->driver = Craft::$app->getConfig()->getDb()->driver;
        if ($this->createTables()) {
            Craft::$app->db->schema->refresh();
        }
        return true;
    }

    /**
     * @return boolean return a false value to indicate the migration fails
     * and should not proceed further. All other return values mean the migration succeeds.
     */
    public function safeDown()
    {
        $this->driver = Craft::$app->getConfig()->getDb()->driver;
        $this->removeTables();
        return true;
    }

    /**
     * Creates the tables needed for the Records used by the plugin
     * @return bool
     */
    protected function createTables()
    {
        $tablesCreated = false;
        $tableSchema = Craft::$app->db->schema->getTableSchema('{{%facebookconnector_accesstoken}}');
        if ($tableSchema === null) {
            $tablesCreated = true;
            $this->createTable(
                '{{%facebookconnector_accesstoken}}',
                [
                    'id' => $this->primaryKey(),
                    'dateCreated' => $this->dateTime()->notNull(),
                    'dateUpdated' => $this->dateTime()->notNull(),
                    'uid' => $this->uid(),
                    'data' => $this->text(),
                ]
            );
        }
        $tableSchema = Craft::$app->db->schema->getTableSchema('{{%facebookconnector_postmemorize}}');
        if ($tableSchema === null) {
            $tablesCreated = true;
            $this->createTable(
                '{{%facebookconnector_postmemorize}}',
                [
                    'id' => $this->primaryKey(),
                    'dateCreated' => $this->dateTime()->notNull(),
                    'dateUpdated' => $this->dateTime()->notNull(),
                    'uid' => $this->uid(),
                    'entryId' => $this->integer()->unique(),
                    'facebookId' => $this->string(255)->unique(),
                    'checksum' => $this->string(255)->notNull()
                ]
            );
        }
        $tableSchema = Craft::$app->db->schema->getTableSchema('{{%facebookconnector_facebook_entry}}');
        if ($tableSchema === null) {
            $tablesCreated = true;
            $this->createTable(
                '{{%facebookconnector_facebook_entry}}',
                [
                    'id' => $this->primaryKey(),
                    'fbId' => $this->string(255)->unique(),
                    'slug' => $this->string(255)->unique(),
                    'dateCreated' => $this->dateTime()->notNull(),
                    'dateUpdated' => $this->dateTime()->notNull(),
                    'created' => $this->string(255)->notNull(),
                    'start_time' => $this->string(45),
                    'end_time' => $this->string(45),
                    'event_cover_source' => $this->string(255),
                    'event_cover_offset_x' => $this->string(45),
                    'event_cover_offset_y' => $this->string(45),
                    'has_detail' => $this->boolean(),
                    'type' => $this->string(),
                    'image_src_1' => $this->string(510),
                    'image_src_2' => $this->string(),
                    'image_src_3' => $this->string(),
                    'image_src_4' => $this->string(),
                    'image_src_5' => $this->string(),
                    'image_src_6' => $this->string(),
                    'image_src' => $this->string(510),
                    'image_height' => $this->string(),
                    'image_width' => $this->string(),
                    'target' => $this->string(510),
                    'url' => $this->string(510),
                    'internal' => $this->boolean(),
                    'uid' => $this->uid(),
                    'title' => $this->string(510),
                    'video' => $this->string(510),
                    'content' => $this->text()
                ]
            );
        }
        return $tablesCreated;
    }

    /**
     * Removes the tables needed for the Records used by the plugin
     * @return void
     */
    protected function removeTables()
    {
        $this->dropTableIfExists('{{%facebookconnector_accesstoken}}');
        $this->dropTableIfExists('{{%facebookconnector_postmemorize}}');
        $this->dropTableIfExists('{{%facebookconnector_facebook_entry}}');
    }
}
